### PR TITLE
Added OptimizeMoleculeConfs with pre-generated force-field

### PR DESCRIPTION
**Pull request by ptosco**
- added two convenience functions to allow miniming all molecule confs with a pre-generated force-field (e.g., to use constraints)
- removed a lot of code duplication between Code/GraphMol/ForceFieldHelpers/MMFF/MMFF.h Code/GraphMol/ForceFieldHelpers/UFF/UFF.h
- homogenized the nonBondedThreshold to 100.0 across all MMFF functions
- added relevant C++ tests

### CODE REVIEW

**Comment by ptosco** [comment]
@greglandrum This is another one for release.

File: Code/GraphMol/ForceFieldHelpers/FFConvenience.h [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    int needsMore = ff.minimize(maxIters);
+    double e = ff.calcEnergy();
+    (*res)[i] = std::make_pair(needsMore, e);
+  }
+}
+
+void OptimizeMoleculeConfsMT(ROMol &mol, const ForceFields::ForceField *ff,
+                               std::vector<std::pair<int, double>> &res,
+                               int numThreads, int maxIters) {
+  std::vector<std::thread> tg;
```
Comment (greglandrum) [inline Code/GraphMol/ForceFieldHelpers/FFConvenience.h]: Need a `PRECONDITION()` on the force field

File: Code/GraphMol/ForceFieldHelpers/FFConvenience.h [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    tg.emplace_back(std::thread(detail::OptimizeMoleculeConfsHelper_,
+                                *ff, &mol, &res, ti, numThreads, maxIters));
+  }
+  for (auto &thread : tg) {
+    if (thread.joinable()) thread.join();
+  }
+}
+#endif
+
+void OptimizeMoleculeConfsST(ROMol &mol, ForceFields::ForceField *ff,
```
Comment (greglandrum) [inline Code/GraphMol/ForceFieldHelpers/FFConvenience.h]: same here with the `PRECONDITION()`

File: Code/GraphMol/ForceFieldHelpers/FFConvenience.h [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  \param ff         the force-field
+  \param res        vector of (needsMore,energy) pairs
+  \param maxIters   the maximum number of force-field iterations
+
+  \return a pair with:
+     first: -1 if parameters were missing, 0 if the optimization converged, 1 if
+  more iterations are required.
+     second: the energy
+*/
+std::pair<int, double> OptimizeMolecule(ForceFields::ForceField *ff, int maxIters = 1000) {
```
Comment (greglandrum) [inline Code/GraphMol/ForceFieldHelpers/FFConvenience.h]: again with the `PRECONDITION()`

File: Code/GraphMol/ForceFieldHelpers/FFConvenience.h [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  \param ff         the force-field
+  \param res        vector of (needsMore,energy) pairs
+  \param numThreads the number of simultaneous threads to use (only has an
+                    effect if the RDKit is compiled with thread support).
+                    If set to zero, the max supported by the system will be
+  used.
+  \param maxIters   the maximum number of force-field iterations
+
+*/
+void OptimizeMoleculeConfs(ROMol &mol, ForceFields::ForceField *ff,
```
Comment (greglandrum) [inline Code/GraphMol/ForceFieldHelpers/FFConvenience.h]: and again

**Comment by ptosco** [comment]
@greglandrum I had indeed forgotten a bunch of files. Even worse, I didn't have them on my own machine anymore either, so I had t redo the whole thing :-/

**Comment by greglandrum** [comment]
Oh crap, that’s awful. Thanks for doing it again.

**Comment by greglandrum** [comment]
Thanks @ptosco!

**Pull request merged**